rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_equipment_associated_with_each_swh_distribution_system.py

```diff
@@ -0,0 +1,85 @@
+from dataclasses import dataclass, field
+from typing import List
+
+from rct229.rulesets.ashrae9012019.ruleset_functions.get_spaces_served_by_swh_use import (
+    get_spaces_served_by_swh_use,
+)
+from rct229.utils.jsonpath_utils import find_all
+
+
+@dataclass
+class SWHDistributionAssociations:
+    swh_heating_eq: List[str] = field(default_factory=list)
+    pumps: List[str] = field(default_factory=list)
+    tanks: List[str] = field(default_factory=list)
+    piping: List[str] = field(default_factory=list)
+    solar_thermal: List[str] = field(default_factory=list)
+    uses: List[str] = field(default_factory=list)
+    spaces_served: List[str] = field(default_factory=list)
+
+
+def get_swh_equipment_associated_with_each_swh_distribution_system(
+    rmd: dict,
+) -> dict[str, SWHDistributionAssociations]:
+    """
+    This function gets all the SWH equipment connected to a SWH distribution system.  The information is stored in a dictionary where the keys are the SWH Distribution System Ids and values are a dictionary giving the ServiceWaterHeatingEquipment, and Pumps connected to the particular SWH distribution system.
+
+    Parameters
+    ----------
+    rmd: dict, RMD at RuleSetModelDescription level
+
+    Returns
+    -------
+    swh_and_equip_dict: A dictionary containing where the keys are the SWH Distribution System IDs and values are dictionaries where keys are the type of SWH equipment and values are the ids of the connected equipment.
+                        Example:  {"swh_distribution1":{"swh_heating_eq":["swh_eq1","swh_eq2"], "pumps":["p1"], "tanks":["t1"], "piping":["piping1"], "solar_thermal":[], "uses":["sp1_use","sp2_use"], "spaces_served":[sp1,sp2]}}
+
+    """
+
+    swh_and_equip_dict = {}
+    for distribution in find_all(
+        "$.service_water_heating_distribution_systems[*]", rmd
+    ):
+        swh_and_equip_dict[distribution["id"]] = SWHDistributionAssociations()
+
+        for tank in distribution.get("tanks", []):
+            swh_and_equip_dict[distribution["id"]].tanks.append(tank["id"])
+
+        for piping in distribution.get("service_water_piping", []):
+            queue = [piping]
+            piping_ids = []
+            # BFS approach
+            while queue:
+                current_piping = queue.pop(0)
```

Comment:
This is small thing - `queue.pop` has a O(n) complexity. I would suggest using `queue.popleft`, O(1) complexity.
However, I do not think this gonna be a huge issue since the set should be small enough to tell the differences.